File: models/commerce.py
from datetime import datetime
from decimal import Decimal
from uuid import uuid4

from sqlalchemy import JSON, BigInteger, Boolean, ForeignKey, Index, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from enums.commerce import PurchaseStatus
from helpers.dates import now
from helpers.db import Base
from helpers.search import search_index
from models.account import Currency
from models.base import AddressedMixin, BigId, IdentifiedMixin, Money, TimestampMixin, UtcDateTime, enum_type, tenant_scoped_unique
from models.tenant import Tenant
from models.user import User


class Product(Base, IdentifiedMixin, AddressedMixin, TimestampMixin):
    """Something bought once and owned for good, which a plan may also hand over as a benefit."""

    __tablename__ = "commerce_product"
    __table_args__ = (UniqueConstraint("uuid", name="commerce_product_uuid"), tenant_scoped_unique("commerce_product_slug", "slug"), Index("commerce_product_listing", "tenant_id", "active", "position"), search_index("commerce_product_search", "name"))

    tenant_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("tenant.id", ondelete="CASCADE"), nullable=True)

    name: Mapped[str] = mapped_column(String(255), nullable=False)
    slug: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    image: Mapped[str | None] = mapped_column(String(512), nullable=True)

    # What owning it lets the account download, which only somebody who owns it is ever handed an address for.
    file: Mapped[str | None] = mapped_column(String(512), nullable=True)

    currency: Mapped[str] = mapped_column(String(3), default="USD", nullable=False)
    price: Mapped[Decimal] = mapped_column(Money, default=Decimal("0"), nullable=False)

    # What owning it puts in a balance, applied the same way whether it was bought or handed over by a plan.
    credits: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    credits_currency_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("currency.id", ondelete="RESTRICT"), nullable=True)

    featured: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    position: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    meta: Mapped[dict] = mapped_column("metadata", JSON, default=dict, nullable=False)

    tenant: Mapped[Tenant | None] = relationship(Tenant)
    credits_currency: Mapped[Currency | None] = relationship(Currency)


class Purchase(Base, IdentifiedMixin, TimestampMixin):
    """One payment for one product, written on this side before the buyer is ever sent to a gateway."""

    __tablename__ = "commerce_purchase"
    __table_args__ = (UniqueConstraint("reference", name="commerce_purchase_reference"), UniqueConstraint("integration_id", "external_id", name="commerce_purchase_integration_external"), Index("commerce_purchase_owner", "user_id", "created_at"), Index("commerce_purchase_tenant_status", "tenant_id", "status"))

    tenant_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("tenant.id", ondelete="CASCADE"), nullable=True)
    user_id: Mapped[int] = mapped_column(BigId, ForeignKey("user.id", ondelete="CASCADE"), nullable=False)
    product_id: Mapped[int] = mapped_column(BigId, ForeignKey("commerce_product.id", ondelete="RESTRICT"), nullable=False)
    integration_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("integration.id", ondelete="RESTRICT"), nullable=True)

    # Minted before the buyer leaves, because what a gateway echoes back has to name a row it never saw being written.
    reference: Mapped[str] = mapped_column(String(64), default=lambda: f"purchase-{uuid4()}", nullable=False)
    external_id: Mapped[str | None] = mapped_column(String(191), nullable=True)

    currency: Mapped[str] = mapped_column(String(3), nullable=False)
    price: Mapped[Decimal] = mapped_column(Money, nullable=False)

    status: Mapped[PurchaseStatus] = mapped_column(enum_type(PurchaseStatus, 16), default=PurchaseStatus.PENDING, nullable=False)
    paid_at: Mapped[datetime | None] = mapped_column(UtcDateTime, nullable=True)
    meta: Mapped[dict] = mapped_column("metadata", JSON, default=dict, nullable=False)

    tenant: Mapped[Tenant] = relationship(Tenant)
    user: Mapped[User] = relationship(User)
    product: Mapped[Product] = relationship(Product)


class UserProduct(Base, IdentifiedMixin, TimestampMixin):
    """What the account owns for good, however it got there, because nothing here ever takes one back."""

    __tablename__ = "user_product"
    __table_args__ = (UniqueConstraint("user_id", "product_id", name="user_product_unique"), Index("user_product_owned", "user_id", "granted_at"))

    user_id: Mapped[int] = mapped_column(BigId, ForeignKey("user.id", ondelete="CASCADE"), nullable=False)
    product_id: Mapped[int] = mapped_column(BigId, ForeignKey("commerce_product.id", ondelete="RESTRICT"), nullable=False)
    purchase_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("commerce_purchase.id", ondelete="SET NULL"), nullable=True)
    subscription_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("subscription.id", ondelete="SET NULL"), nullable=True)
    benefit_grant_id: Mapped[int | None] = mapped_column(BigId, ForeignKey("subscription_benefit_grant.id", ondelete="SET NULL"), nullable=True)

    granted_at: Mapped[datetime] = mapped_column(UtcDateTime, default=now, nullable=False)
    meta: Mapped[dict] = mapped_column("metadata", JSON, default=dict, nullable=False)

    user: Mapped[User] = relationship(User)
    product: Mapped[Product] = relationship(Product)

File: webapps/admin/src/resources/commerce.js
import { auditGroup, choice, decimal, file, html, image, lookup, metadata, number, tenantField, text, toggle } from "./fields";

export const products = {
    name: "products",
    ordering: ["id", "name", "slug", "price", "position", "createdAt"],
    defaultOrdering: "position",
    section: "commerce",
    icon: "tag",
    labelField: "name",
    columns: [
        { name: "image", label: "field.image", type: "thumbnail" },
        { name: "name", label: "field.name" },
        { name: "slug", label: "field.slug", type: "code" },
        { name: "tenant", label: "field.tenant", type: "reference", referenceField: "name" },
        { name: "price", label: "field.price", type: "number" },
        { name: "credits", label: "field.credits", type: "number" },
        { name: "active", label: "field.active", type: "boolean" },
    ],
    filters: [
        { name: "tenantId", label: "field.tenant", type: "lookup", resource: "tenants" },
        { name: "featured", label: "field.featured", type: "boolean" },
        { name: "active", label: "field.active", type: "boolean" },
    ],
    groups: [
        { key: "identification", fields: [text("name", "field.name", { required: true }), text("slug", "field.slug"), tenantField] },
        { key: "content", fields: [html("description", "field.description"), image("image", "field.image", "product-image"), file("file", "field.file", "product-file", { hint: "field.productFileHint" })] },
        { key: "pricing", fields: [text("currency", "field.currency", { required: true, default: "USD" }), decimal("price", "field.price", { min: 0, default: 0 })] },
        { key: "credits", fields: [number("credits", "field.credits", { min: 0, default: 0, hint: "field.creditsHint" }), lookup("creditsCurrencyId", "field.currency", "currencies")] },
        { key: "presentation", fields: [toggle("featured", "field.featured"), number("position", "field.position", { min: 0, default: 0 })] },
        { key: "advanced", fields: [toggle("active", "field.active", { default: true }), metadata] },
        auditGroup,
    ],
};

export const purchases = {
    name: "purchases",
    ordering: ["id", "status", "price", "paidAt", "createdAt"],
    section: "commerce",
    icon: "card",
    labelField: "reference",
    readOnly: true,
    columns: [
        { name: "reference", label: "field.reference", type: "code" },
        { name: "user", label: "field.user", type: "reference", referenceField: "displayName" },
        { name: "product", label: "field.product", type: "reference", referenceField: "name" },
        { name: "price", label: "field.price", type: "number" },
        { name: "status", label: "field.status", type: "enum", enumName: "purchase_status" },
        { name: "paidAt", label: "field.paidAt", type: "datetime" },
    ],
    filters: [
        { name: "tenantId", label: "field.tenant", type: "lookup", resource: "tenants" },
        { name: "userId", label: "field.user", type: "lookup", resource: "users" },
        { name: "productId", label: "field.product", type: "lookup", resource: "products" },
        { name: "integrationId", label: "field.integration", type: "lookup", resource: "integrations" },
        { name: "status", label: "field.status", type: "enum", enumName: "purchase_status" },
    ],
    viewExtra: [
        { name: "tenant", label: "field.tenant", type: "reference", referenceField: "name" },
        { name: "externalId", label: "field.externalId", type: "code" },
        { name: "currency", label: "field.currency", type: "code" },
        { name: "meta", label: "field.metadata", type: "json" },
        { name: "createdAt", label: "field.createdAt", type: "datetime" },
        { name: "updatedAt", label: "field.updatedAt", type: "datetime" },
    ],
};

export const userProducts = {
    name: "user-products",
    searchable: false,
    ordering: ["id", "grantedAt", "createdAt"],
    section: "commerce",
    icon: "gift",
    labelField: "id",
    readOnly: true,
    columns: [
        { name: "user", label: "field.user", type: "reference", referenceField: "displayName" },
        { name: "product", label: "field.product", type: "reference", referenceField: "name" },
        { name: "grantedAt", label: "field.grantedAt", type: "datetime" },
    ],
    filters: [
        { name: "userId", label: "field.user", type: "lookup", resource: "users" },
        { name: "productId", label: "field.product", type: "lookup", resource: "products" },
        { name: "subscriptionId", label: "field.subscription", type: "lookup", resource: "subscriptions" },
    ],
    viewExtra: [
        { name: "purchaseId", label: "field.purchase", type: "number" },
        { name: "subscriptionId", label: "field.subscription", type: "number" },
        { name: "benefitGrantId", label: "field.benefitGrant", type: "number" },
        { name: "meta", label: "field.metadata", type: "json" },
        { name: "createdAt", label: "field.createdAt", type: "datetime" },
    ],
};
